### PR TITLE
atomic registry: projects selector

### PR DESCRIPTION
**Pull request by sub-mod**

### CODE REVIEW

File: pkg/registry/scripts/projects.js [hunk 1]
```diff
@@ -30,11 +30,17 @@
...
-            $routeProvider.when('/projects/:namespace?', {
-                templateUrl: 'views/projects-page.html',
-                controller: 'ProjectsCtrl',
-                reloadOnSearch: false,
-            });
+            $routeProvider
+                .when('/projects/:namespace?', {
+                    controller: 'ProjectsCtrl',
+                    templateUrl: function(params) {
+                        var namespace = params['namespace'] || '';
```
Comment (petervo) [inline pkg/registry/scripts/projects.js]: Is there a reason not to do if (!params['namespace']) here?

File: pkg/registry/scripts/projects.js [hunk 2]
```diff
@@ -85,33 +91,88 @@
...
-        function($scope, loader, select, $modal, util, projectAction) {
-            loader.watch("users");
-            loader.watch("groups");
-            loader.load("Project", null, null);
+        '$routeParams',
+        'ListingState',
+        '$location',
+        function($scope, loader, select, $modal, util, projectAction, $routeParams, ListingState, $location) {
+            var namespace = $routeParams["namespace"] || "";
+            if(namespace!== "") {
```
Comment (petervo) [inline pkg/registry/scripts/projects.js]: spacing, also what about the no namespace case, should we redirect somewhere?

File: pkg/registry/views/project-listing.html [hunk 3]
```diff
@@ -0,0 +1,78 @@
+<table listing-table class="listing projects-listing">
+    <thead >
+        <tr ng-if="!listing.single">
+            <td colspan="3">
+                <a class="pull-right" ng-click="createProject()">
+                    <i translatable="yes" class="pficon pficon-add-circle-o"></i>
+                    <span translatable="yes">New Project</span>
+                </a>
+                <h3 translatable="yes" ng-click="console.log(projects())">Projects</h3>
```
Comment (petervo) [inline pkg/registry/views/project-listing.html]: console.log?

File: pkg/registry/views/project-panel.html [hunk 4]
```diff
@@ -0,0 +1,39 @@
+<div class="listing-head" ng-click="listing.collapse(id, $event)">
+    <div class="listing-actions">
+        <button class="btn btn-danger btn-delete pficon pficon-delete" disabled></button>
+        <button class="btn btn-default pficon pficon-edit" disabled></button>
+    </div>
+    <h3><i class="pficon pficon-project"></i>{{ project.metadata.name }}</h3>
+    <h4>{{ project.metadata.annotations["openshift.io/description"]}}</h4>
+</div>
+<!-- <div project-body class="listing-body project-panel"  ng-show="tab('main')"></div> -->
```
Comment (petervo) [inline pkg/registry/views/project-panel.html]: Still needed?

File: pkg/registry/views/project-panel.html [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+<div class="listing-body project-panel" ng-show="tab('main')">
+    <div>
+        <table listing-table class="listing project-body">
+            <thead >
+                <tr >
+                    <th translatable="yes">Project Members</th>
+                    <th translatable="yes">Roles</th>
+                    <th translatable="yes"></th>
```
Comment (petervo) [inline pkg/registry/views/project-panel.html]: Do we need the extra rows?

File: pkg/registry/views/project-panel.html [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+                </tr>
+            </thead>
+            <tbody colspan="3">
+                <tr class="listing-item inner-project-listing ">
+                    <td translatable="yes">todo members</td>
```
Comment (petervo) [inline pkg/registry/views/project-panel.html]: todos shouldn't be translated

**Comment by petervo** [comment]
I didn't comment on each instance but a little more spacing consistency would be nice in the templates with {{ and function calls.

File: pkg/registry/views/project-panel.html (hunk 5 above)
Comment (sub-mod) [inline pkg/registry/views/project-panel.html]: yes

Comment (petervo) [inline pkg/registry/views/project-panel.html]: what is it used for?

Comment (sub-mod) [inline pkg/registry/views/project-panel.html]: there is a delete button coming here

**Comment by petervo** [comment]
Deleting a project doesn't remove it from the list.

**Comment by stefwalter** [comment]
This commit needs a properly descriptive commit message stating what is going on here.

**Comment by sub-mod** [comment]
@petervo I am not able to see this issue. I have deleted a project and I am able to see that the project is not listed anymore on the UI.

**Comment by sub-mod** [comment]
@stefwalter done

File: pkg/registry/scripts/projects.js [hunk 7]
```diff
@@ -85,33 +90,88 @@
...
-            $scope.groups = function() {
-                return select().kind("Group");
-            };
+            } else {
 
-            $scope.projects = function() {
-                return select().kind("Project");
-            };
+                $scope.listing = new ListingState($scope);
+                $scope.listing.single = false;
```
Comment (stefwalter) [inline pkg/registry/scripts/projects.js]: Why is this set to false?

In addition the menu (upper right) has an 'Expand items inline' option. It has no effect. It should have an effect, and currently the default would be to not 'expand items inline' (ie: menu option not checked)

File: pkg/registry/scripts/projects.js [hunk 8]
```diff
@@ -85,33 +90,88 @@
...
+                link: function(scope, element, attrs) {
+                    var tab = 'main';
+                    scope.tab = function(name, ev) {
+                        if (ev) {
+                            tab = name;
+                            ev.stopPropagation();
+                        }
+                        return tab === name;
+                    };
+                    //get policy stuff here
```
Comment (stefwalter) [inline pkg/registry/scripts/projects.js]: Is this a TODO? Suggest adding that notation if so.

File: pkg/registry/scripts/projects.js [hunk 9]
```diff
@@ -85,33 +90,88 @@
...
+                    var currProject = scope.id;
+                    loader.load("Project", null, null);
+                    scope.project = select().kind("Project").name(currProject).one();
```
Comment (stefwalter) [inline pkg/registry/scripts/projects.js]: How does this get updated when/if the project changes?

File: pkg/registry/styles/projects.less [hunk 10]
```diff
@@ -1,8 +1,52 @@
...
+.project-panel {
+    padding-right: 25px;
+    padding-top: 0px;
+}
+
 table.projects-listing {
+    max-width: 2000px;
+    min-width: 85%;
+
+    tbody:hover {
```
Comment (stefwalter) [inline pkg/registry/styles/projects.less]: Is this really necessary? What effect does it have? Removing it seems to have no effect.

File: pkg/registry/styles/projects.less [hunk 11]
```diff
@@ -1,8 +1,52 @@
...
+        background-color: #d4edfa;
+    }
+
+    tr.project-item {
```
Comment (stefwalter) [inline pkg/registry/styles/projects.less]: There is no mention of 'project-item' anywhehre outside this less file. I think this block is unnecessary.

File: pkg/registry/styles/projects.less [hunk 12]
```diff
@@ -19,6 +63,15 @@ table.projects-listing {
         cursor: pointer;
     }
 
+    tbody.inner-project-listing-add {
```
Comment (stefwalter) [inline pkg/registry/styles/projects.less]: I see no mention of 'inner-project-listing-add' outside of this less file. I think this change is unnecessary.

File: pkg/registry/styles/projects.less [hunk 13]
```diff
@@ -19,6 +63,15 @@ table.projects-listing {
...
+        border-bottom: 0px;
+    }
+
+    tr.inner-project-listing {
+        display:block;
+        display: table-row;
```
Comment (stefwalter) [inline pkg/registry/styles/projects.less]: There are two conflicting display attributes. Could you what this does and why it's necessary?

File: pkg/registry/views/project-listing.html [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
+        </tr>
+    </thead>
+    <tbody ng-repeat-start="item in projects() track by item.metadata.name"  ng-init="sid = item.metadata.name" 
+           data-id="{{ sid }}" class="active" ng-class="{ open: listing.expanded(sid) }">
+        <tr class="listing-item" ng-click="listing.toggle(sid)">
+            <th>{{ item.metadata.name }}</th>
+            <td>{{ item.metadata.annotations["openshift.io/description"] }}</td>
+            <td></td>
+        </tr>
+        <tr class="listing-panel" ng-if="listing.expanded(sid )" ng-init="id = sid">
```
Comment (stefwalter) [inline pkg/registry/views/project-listing.html]: Extra space after sid here.

File: pkg/registry/scripts/projects.js (hunk 7 above)
Comment (sub-mod) [inline pkg/registry/scripts/projects.js]: How do I expand all items ? is there an example else where

File: pkg/registry/scripts/projects.js (hunk 9 above)
Comment (sub-mod) [inline pkg/registry/scripts/projects.js]: I am not sure how to do that.Is there an example someplace for this?

Comment (sub-mod) [inline pkg/registry/scripts/projects.js]: ok I wrapped it in a function now.

File: pkg/registry/scripts/projects.js (hunk 7 above)
Comment (sub-mod) [inline pkg/registry/scripts/projects.js]: ok works now

File: pkg/registry/scripts/projects.js [hunk 15]
```diff
@@ -85,33 +90,94 @@
...
+
+                $scope.createProject = projectAction.createProject;
+
+                $scope.createGroup = projectAction.createGroup;
 
-            $scope.formatMembers = util.formatMembers;
+                $scope.createUser = projectAction.createUser;
 
-            $scope.createProject = projectAction.createProject;
+                $scope.$on("activate", function(ev, id) {
```
Comment (sub-mod) [inline pkg/registry/scripts/projects.js]: added code for  per item view OR inline view

File: pkg/registry/scripts/projects.js [hunk 16]
```diff
@@ -85,33 +90,94 @@
...
+                            tab = name;
+                            ev.stopPropagation();
+                        }
+                        return tab === name;
+                    };
 
-            $scope.createUser = projectAction.createUser;
+                    var currProject = scope.id;
+                    loader.load("Project", null, null);
+                    scope.project = function() {
```
Comment (sub-mod) [inline pkg/registry/scripts/projects.js]: added code to track project changes

**Comment by stefwalter** [comment]
Looks good. Waiting for tests.

**Pull request closed without merging**